go/test/endtoend/vreplication/vreplication_test.go

```diff
@@ -826,6 +826,107 @@ func shardCustomer(t *testing.T, testReverse bool, cells []*Cell, sourceCellOrAl
 			shardNames = append(shardNames, shardName)
 		}
 		testSwitchTrafficPermissionChecks(t, workflowType, sourceKs, shardNames, targetKs, workflow)
+
+		// Confirm that switching writes works as expected in the face of
+		// vreplication lag (canSwitch() precheck) and when cancelling the
+		// switch due to replication failing to catch up in time.
+		t.Run("validate switch writes error handling", func(t *testing.T) {
```

Comment:
let's extract this into a separate method.

Reply:
It depends on the state there so I don't really see the value in moving it out and passing all of those things in (source and target connections, queries, etc.) as it will still be dependent on the state setup in shardCustomer.

Reply:
Looks like most of the vars used (e.g. productTab) are globals, so if you feel strongly about it I can move it to its own function.

Reply:
I did end up moving it to its own function and generalizing it so that we can also test Reshard in https://github.com/vitessio/vitess/pull/17616/commits/76396decac35fdfabd9c707ecb4c5c718771981a